frontend/src/metabase/querying/utils/drills/combine-columns-drill/components/CombineColumnsDrill/CombineColumnsDrill.tsx

```diff
@@ -0,0 +1,185 @@
+      { ...value[index], ...change },
+      ...value.slice(index + 1),
+    ]);
+  };
+
+  const handleAdd = () => {
+    setColumnsAndSeparators(value => [
+      ...value,
+      getNextColumnAndSeparator(
+        expressionableColumns,
+        defaultSeparator,
+        options,
+        columnsAndSeparators,
+      ),
+    ]);
+  };
+
+  const handleRemove = (index: number) => {
+    setColumnsAndSeparators(value => [
+      ...value.slice(0, index),
+      ...value.slice(index + 1),
+    ]);
+  };
+
+  const handleEditSeparators = () => {
+    setIsUsingDefaultSeparator(false);
+  };
+
+  const handleSubmit: FormEventHandler = event => {
+    event.preventDefault();
+
+    const name = getExpressionName(
+      query,
+      stageIndex,
+      column,
+      columnsAndSeparators,
+    );
+    const newQuery = Lib.expression(query, stageIndex, name, expressionClause);
+
+    onSubmit(newQuery);
+  };
+
+  return (
+    <form onSubmit={handleSubmit}>
+      <Card maw="100vw" w={474} p="lg">
+        <Title
+          mb="lg"
+          order={4}
+        >{jt`Combine “${columnInfo.displayName}” with other columns`}</Title>
+
+        <Stack spacing="lg">
+          <Stack spacing={12}>
+            <ScrollArea mx={-OVERFLOW_SAFETY_MARGIN}>
+              <Box mah="50vh" px={OVERFLOW_SAFETY_MARGIN}>
+                <Stack spacing="sm">
+                  {columnsAndSeparators.map(({ column, separator }, index) => (
+                    <ColumnAndSeparatorRow
+                      column={column}
+                      index={index}
+                      key={index}
+                      options={options}
+                      separator={separator}
+                      showLabels={index === 0}
+                      showRemove={columnsAndSeparators.length > 1}
+                      showSeparator={!isUsingDefaultSeparator}
+                      onChange={handleChange}
+                      onRemove={handleRemove}
+                    />
+                  ))}
+                </Stack>
+              </Box>
+            </ScrollArea>
+
+            <Flex
+              align="center"
+              gap="md"
+              justify={isUsingDefaultSeparator ? "space-between" : "end"}
+            >
+              {isUsingDefaultSeparator && (
+                <Box>
+                  <Button p={0} variant="subtle" onClick={handleEditSeparators}>
+                    {jt`Separated by ${formatSeparator(defaultSeparator)}`}
+                  </Button>
+                </Box>
+              )}
+
+              <Button
+                leftIcon={<Icon name="add" />}
+                p={0}
+                variant="subtle"
+                onClick={handleAdd}
+              >
+                {t`Add column`}
+              </Button>
+            </Flex>
+          </Stack>
+
+          <Example example={example} />
+
+          <Flex align="center" gap="md" justify="end">
+            <Button type="submit" variant="filled">
```

Comment:
We should probably disable the button when the form is in an invalid state.

Reply:
I don't think this form can be in an invalid state. Am I missing something?

Reply:
I guess it's a decision we can make, but when a column dropdown I think the empty the form is in an invalid state. The alternative is that we just ignore that column in the resulting expression.

Reply:
In this branch column dropdown is never empty.
When adding a new "row", a column is automatically selected, see `getNextColumnAndSeparator` and how it's used.